tests/posix/common/src/clock.c

```diff
@@ -8,118 +9,207 @@
+static const struct timespec ref_ts = {1514821501, NSEC_PER_SEC / 2U};
 
-	/* TESTPOINT: Pass invalid clock type */
-	zassert_equal(clock_gettime(CLOCK_INVALID, &ts), -1,
-			NULL);
-	zassert_equal(errno, EINVAL);
+static const clockid_t clocks[] = {
+	CLOCK_MONOTONIC,
+	CLOCK_REALTIME,
+};
+static const bool settable[] = {
+	false,
+	true,
+};
+
+static inline int64_t ts_to_ns(const struct timespec *ts)
+{
+	return ts->tv_sec * NSEC_PER_SEC + ts->tv_nsec;
+}
+
+static inline void tv_to_ts(const struct timeval *tv, struct timespec *ts)
+{
+	ts->tv_sec = tv->tv_sec;
+	ts->tv_nsec = tv->tv_usec * NSEC_PER_USEC;
+}
 
-	zassert_ok(clock_gettime(CLOCK_MONOTONIC, &ts));
-	zassert_ok(k_sleep(K_SECONDS(SLEEP_SECONDS)));
-	zassert_ok(clock_gettime(CLOCK_MONOTONIC, &te));
+#define _tp_op(_a, _b, _op) (ts_to_ns(_a) _op ts_to_ns(_b))
 
-	if (te.tv_nsec >= ts.tv_nsec) {
-		secs_elapsed = te.tv_sec - ts.tv_sec;
-		nsecs_elapsed = te.tv_nsec - ts.tv_nsec;
-	} else {
-		nsecs_elapsed = NSEC_PER_SEC + te.tv_nsec - ts.tv_nsec;
-		secs_elapsed = (te.tv_sec - ts.tv_sec - 1);
+#define _decl_op(_type, _name, _op)                                                                \
+	static inline _type _name(const struct timespec *_a, const struct timespec *_b)            \
+	{                                                                                          \
+		return _tp_op(_a, _b, _op);                                                        \
 	}
 
-	/*TESTPOINT: Check if POSIX clock API test passes*/
-	zassert_equal(secs_elapsed, SLEEP_SECONDS,
-			"POSIX clock API test failed");
+_decl_op(bool, tp_eq, ==);     /* a == b */
+_decl_op(bool, tp_lt, <);      /* a < b */
+_decl_op(bool, tp_gt, >);      /* a > b */
+_decl_op(bool, tp_le, <=);     /* a <= b */
+_decl_op(bool, tp_ge, >=);     /* a >= b */
+_decl_op(int64_t, tp_diff, -); /* a - b */
+
+/* lo <= (a - b) < hi */
+static inline bool tp_diff_in_range_ns(const struct timespec *a, const struct timespec *b,
+				       int64_t lo, int64_t hi)
+{
+	int64_t diff = tp_diff(a, b);
 
-	printk("POSIX clock APIs test done\n");
+	return diff >= lo && diff < hi;
 }
 
-ZTEST(posix_apis, test_realtime)
+ZTEST(posix_apis, test_clock_gettime)
 {
-	int ret;
-	struct timespec rts, mts;
-	struct timeval tv;
+	struct timespec ts;
 
-	ret = clock_gettime(CLOCK_MONOTONIC, &mts);
-	zassert_equal(ret, 0, "Fail to get monotonic clock");
-
-	ret = clock_gettime(CLOCK_REALTIME, &rts);
-	zassert_equal(ret, 0, "Fail to get realtime clock");
+	/* ensure argument validation is performed */
+	errno = 0;
+	zassert_equal(clock_gettime(CLOCK_INVALID, &ts), -1);
+	zassert_equal(errno, EINVAL);
 
-	/* Set a particular time.  In this case, the output of:
-	 * `date +%s -d 2018-01-01T15:45:01Z`
-	 */
-	struct timespec nts;
-	nts.tv_sec = 1514821501;
-	nts.tv_nsec = NSEC_PER_SEC / 2U;
+	if (false) {
+		/* not hardened */
+		errno = 0;
+		zassert_equal(clock_gettime(clocks[0], NULL), -1);
+		zassert_equal(errno, EINVAL);
+	}
```

Comment:
What is the purpose of this?

Reply:
It's there because that is technically undefined behaviour. The behaviour as it's been for quite a long time has been to fault.

Reply:
So why not just remove it?

There is not much point in keeping it if it is going to be permanently disabled. It would make more sense if there were a Kconfig to optionally enable this test; but, that would be quite unnecessary.

Reply:
> There is not much point in keeping it if it is going to be permanently disabled.

There is a point. 

It shows that the scenario was considered in the design of the testsuite but is intentionally disabled.

> It would make more sense if there were a Kconfig to optionally enable this test; but, that would be quite unnecessary.

That would actually make significantly less sense.

Reply:
> It shows that the scenario was considered in the design of the testsuite but is intentionally disabled. That would provide a hint to any future maintainer of this testsuite that there is something exceptional about this scenario which would allow that person to save some valuable time.

Can you please add a brief comment about this? Because the current comment saying "not hardened" is rather cryptic and I am sure anyone else reading the code in the future will be asking the same question as me.

Reply:
Fixed!